File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/MSCertificateTemplateV2AttributeInfo.kt
package com.jetbrains.signatureverifier.serialization

import kotlinx.serialization.Serializable
import org.bouncycastle.asn1.DLSequence
import org.bouncycastle.asn1.DLTaggedObject
import org.bouncycastle.asn1.cms.Attribute

// 1.3.6.1.4.1.311.2.1.12
@Serializable
data class MSCertificateTemplateV2AttributeInfo(
  val identifier: TextualInfo,
  val content: List<List<TaggedObjectInfo>>
) : AttributeInfo {

  constructor(attribute: Attribute) : this(
    TextualInfo.getInstance(attribute.attrType),
    attribute.attributeValues.map { attributeValue ->
      (attributeValue as DLSequence).map {
        (it as DLTaggedObject).let { outer ->
          TaggedObjectInfo(
            TaggedObjectMetaInfo(outer),
            (outer.baseObject as DLTaggedObject).let { inner ->
              TaggedObjectInfo(
                TaggedObjectMetaInfo(inner),
                TextualInfo.getInstance(inner.baseObject)
              )
            }
          )
        }
      }
    }
  )

  override fun toAttributeDLSequence(): DLSequence =
    listOf(
      identifier.toPrimitive(),
      content.map {
        it.toPrimitiveList().toDLSequence()
      }.toDLSet()
    ).toDLSequence()
}

File: net/JetBrains.Serialization/src/Dataholders/MSCertificateTemplateV2AttributeInfo.cs
using Newtonsoft.Json;
using Org.BouncyCastle.Asn1;
using Attribute = Org.BouncyCastle.Asn1.Cms.Attribute;
using JetBrains.Serialization;
using System.Collections.Generic;
using System.Linq;

[JsonObject(MemberSerialization.Fields)]
public class MSCertificateTemplateV2AttributeInfo : AttributeInfo
{
    public override TextualInfo Identifier { get; }
    public List<List<TaggedObjectInfo>> Content { get; }

    public MSCertificateTemplateV2AttributeInfo(Attribute attribute)
    {
        Identifier = TextualInfo.GetInstance(attribute.AttrType);
        Content = attribute.AttrValues.ToArray().OfType<Asn1Sequence>().Select(seq =>
            seq.OfType<DerTaggedObject>()
               .Select(outer => new TaggedObjectInfo(outer.IsExplicit(), outer.TagNo, new TaggedObjectInfo(
                    ((DerTaggedObject)outer.GetObject()).IsExplicit(),
                    ((DerTaggedObject)outer.GetObject()).TagNo,
                    TextualInfo.GetInstance(((DerTaggedObject)outer.GetObject()).GetObject()))))
               .ToList())
           .ToList();
    }

    public override Asn1Encodable GetPrimitiveContent() => Content.Select(list => list.ToPrimitiveDerSequence()).ToDerSet();
}